### ISSUE TITLE
Add a project type that uses `[dependency-groups]` only and doesn't define a `[project]` table? Sync `dependency-groups.main` by default?

### ISSUE THREAD

**Issue by matterhorn103** [astral-sh/uv#8582]
Amazing to see PEP-735 so fast! And cool of you to adopt it immediately for dev dependencies. You are doing great things 🙂 

One of the motivating use-cases for the PEP was science/data science projects that need to specify dependencies but aren't packages. One reason being that a `[project]` table needs a name and a version which can is a bit weird in that context. Another was the attitude of some of the Python devs that `project.dependencies]` should only be used by things that get packaged and not as an alternative to `requirements.txt`.

The PEP [envisages](https://peps.python.org/pep-0735/#data-science-projects) for this use-case that the `pyproject.toml` file wouldn't contain a `[project]` table at all but instead just a `[dependency-groups]` one containing a `main` group.

Have you guys considered or do you already have plans to support this exact scenario? By this I mean doing the following:

1. Syncing a `dependency-groups.main` by default if present, i.e. special casing it in the same way as you have decided to do for `dependency-groups.dev` (which is a nice move by the way)

2. Adding a dependency to `dependency-groups.main` if `uv add` is run on a project whose `pyproject.toml` doesn't contain a `[project]` table

3. Creating a `pyproject.toml` that matches this setup for some new kind of project distinct from `--app`, `--lib`, and `--package`

This new project type would also be very much applicable to e.g. a folder of random scripts where the user wants to maintain a common set of installed packages and nothing more (this is something I'd use it for, on top of for scientific projects). At the moment this is still best managed  using `uv venv` and `uv pip`.

I don't know what this new project type would be called. "Workspace" would have been fitting but now has a different meaning. "Environment" is honestly not wrong as a way of describing the concept but has so much baggage (`uvenv` anybody? Lol). Maybe a "directory". My best suggestion would be "minimal" or "bare", the latter of which I particularly like. But no idea really.

The closest project type currently is `--app` but neither its behaviour nor its name quite feel like a good fit for the science use-case or the other I mentioned.  So this would help with that too perhaps.

If you were to implement something like this, I also think it'd be worth making "bare" the default project type of `uv init` without flags, as I think the people who are most likely to want this kind of minimal behaviour are also probably the ones least knowledgeable about Python project management. But that's a separate point tbh.

**Comment by zanieb** [comment]
I'll return to this, but in brief we're pretty interested in supporting this use case and, more generally, we'll be able to support dependency groups without a `[project]` section soon.

**Comment by vlcinsky** [comment]
The PEP735 names this type of project "non-package". But this sounds too close to already taken "--no-package" option.

I like the `--bare` proposal.

Btw: I am amazed by the speed it resolves, installs and syncs the packages. It is so fast it is hard to believe it did anything.

**Comment by charliermarsh** [comment]
You can already do something like this with:

```toml
[dependency-groups]
test = ["pytest"]

[tool.uv.workspace]  # Must be present, but can be empty.
```

But that structure (no `[project]` table) is considered legacy in uv right now.

**Comment by bluss** [comment]
PEP735 talks about non-package, but there's no way to actually designate something as that, is there? (In PEP terms). (Mainly thinking about how no build-system == build me with setuptools, so everything with a pyproject.toml becomes a buildable.)

**Comment by zanieb** [comment]
As far as the PEP is concerned, you designate something as non-packaged by omitting a `[project]` table entirely.

**Comment by CarrotManMatt** [comment]
The ability to have a `pyproject.toml` file without a `[project]` section is now possible since uv 0.8.18 and the merge of #14113. This does not, however support locking an environment without a project and requires the use of `managed = false`.

**Comment by zanieb** [comment]
Oh, `uv lock` should support that imo. I think that might just be an oversight as it's not in the list at https://github.com/astral-sh/uv/pull/14113#issuecomment-2981324244. cc @Gankra ?

**Comment by matterhorn103** [comment]
I've only just seen #14113, it's really nice to see you guys implement better support for "project-less" projects!

As far as I can tell, other than the mentioned `uv lock`, the one other thing that hasn't actually been implemented by #14113 is a way to actually create a project with `uv init` that is missing a `[project]` table? While `uv init --bare` was added a while back by #11192 (and I have indeed used it several times), it still creates a `[project]` table.

I'm curious to know if you guys have had any further ideas or concrete plans for the non-versioned/scientific project type. The "Note" box at the top of [the dependency docs page](https://docs.astral.sh/uv/concepts/projects/dependencies/) reads a little like you're waiting to see how adoption of `dependency-groups` pans out and whether the community instead just decides to ignore the original pyproject PEP and co-opt `[project.dependencies]` for non-packages too?

**Comment by CarrotManMatt** [comment]
Looks like support for `uv lock` is now available for `[project]`-less projects in #19087! As @matterhorn103 mentioned, this means that the only remaining problems to resolve this issue are the ability for a `uv init` project type that does not include a `project` table and possibly the decision to sync `dependency-groups.main` (or other bikeshedded spellings) by default for `[project]`-less projects.

**Comment by CarrotManMatt** [comment]
In addition, the way to describe `requires-python` without a `[project]` table is still unresolved.

**Comment by ches** [comment]
> In addition, the way to describe `requires-python` without a `[project]` table is still unresolved.

Just to reference: #13735

I haven't taken more than a minute to examine how this all interacts now that `[project]`-less parts landed, but seems safe to say "not quite what we'd hope" yet for this use case:

```
% uv --version
uv 0.12.1 (Homebrew 2026-07-31 aarch64-apple-darwin)

% tree
.
└── pyproject.toml

1 directory, 1 file

% cat pyproject.toml
[dependency-groups]
dev = ['click']

[tool.uv.dependency-groups]
dev = { requires-python = "==3.13.14" }

% uv run --only-dev python -V
Using CPython 3.14.6
Creating virtual environment at: .venv
warning: No `requires-python` value found in the workspace. Defaulting to `>=3.14`.
Installed 1 package in 5ms
Python 3.14.6
```

**Edit**:

In its original `[project]` context it was perhaps fair to say that `requires-python` *is fundamentally a packaging (metadata) matter* and that `.python-version` is the thing to use for our non-package type use cases that are the subject of this issue.

In solving the real resolution problems that #13735 addressed, the term `requires-python` also got attached to a context that is explicitly _not_ for packaging, namely dependency groups. Perhaps that muddies the waters a little now that "non-package projects" are (almost?) a first-class thing 😅 

The above "looks" like it should do what you'd expect, even if it's a bit of a roundabout/non-intuitive way to spell it for the common case of non-package projects that will only want one group, like `main`. They could use `.python-version` instead, but the above form should probably either work as expected or give a diagnostic.

I can readily imagine non-package projects having uses for multiple dependency groups, like a backend server that wants to exclude its `dev` tool and `test` groups when building environments for deployment, container images, etc.

But needing disjoint Python version requirements in those cases seems a stretch to me, even an anti-pattern. It might be a sign that workspace members would be more appropriate than groups.

Anyway, perhaps this specific matter should be taken to a new issue, if there isn't one already.

**Comment by CarrotManMatt** [comment]
Agreed that `.python-version` can work fine for a `>=` Python version constraint, but it does not allow for any other constrain types. For example in our use-case, we need to pin to `requires-python = "==3.13.*"` without a `[project]` table.

Even if the problem with `[tool.uv.dependency-groups].*.requires-python` is resolved, it would be very unergonomic for us to have to spell out the Python version constraint for each one of our dependency groups separately (we have 6 groups including, like you said, one for deployment installations and one with developer tools).
